presto-main/src/main/java/com/facebook/presto/server/TaskInfoResource.java

```diff
@@ -13,39 +13,96 @@
 import javax.ws.rs.NotFoundException;
 import javax.ws.rs.Path;
 import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.container.AsyncResponse;
+import javax.ws.rs.container.Suspended;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.UriBuilder;
+import javax.ws.rs.core.UriInfo;
 
+import java.net.URI;
+import java.net.UnknownHostException;
 import java.util.Optional;
 
+import static com.google.common.base.Preconditions.checkState;
 import static java.util.Objects.requireNonNull;
+import static javax.ws.rs.core.Response.Status.NOT_FOUND;
 
 @Path("v1/taskInfo")
 @RolesAllowed("ADMIN")
 public class TaskInfoResource
 {
+    public static final String INCLUDE_LOCAL_QUERY_ONLY = "includeLocalQueryOnly";
+    private final DispatchManager dispatchManager;
     private final QueryManager queryManager;
+    private final boolean resourceManagerEnabled;
+    private final InternalNodeManager internalNodeManager;
+    private final Optional<ResourceManagerProxy> proxyHelper;
 
     @Inject
-    public TaskInfoResource(QueryManager queryManager)
+    public TaskInfoResource(
+            DispatchManager dispatchManager,
+            QueryManager queryManager,
+            InternalNodeManager internalNodeManager,
+            ServerConfig serverConfig,
+            Optional<ResourceManagerProxy> proxyHelper)
     {
+        this.dispatchManager = requireNonNull(dispatchManager, "dispatchManager is null");
         this.queryManager = requireNonNull(queryManager, "queryManager is null");
+        this.internalNodeManager = requireNonNull(internalNodeManager, "internalNodeManager is null");
+        this.resourceManagerEnabled = requireNonNull(serverConfig, "serverConfig is null").isResourceManagerEnabled();
+        this.proxyHelper = requireNonNull(proxyHelper, "proxyHelper is null");
     }
 
     @GET
     @Path("{taskId}")
-    public TaskInfo getTaskInfo(@PathParam("taskId") TaskId taskId)
-            throws NotFoundException
+    @Produces(MediaType.APPLICATION_JSON)
+    public void getTaskInfo(@PathParam("taskId") TaskId taskId,
+            @QueryParam(INCLUDE_LOCAL_QUERY_ONLY) @DefaultValue("false") boolean includeLocalQueryOnly,
+            @Context UriInfo uriInfo,
+            @Context HttpServletRequest servletRequest,
+            @Suspended AsyncResponse asyncResponse)
+            throws WebApplicationException
+    {
+        if (requestNeedsToBeProxied(taskId, includeLocalQueryOnly)) {
+            proxyTaskInfo(servletRequest, asyncResponse, uriInfo);
+        }
+        else {
+            try {
+                asyncResponse.resume(Response.ok(getTaskInfo(taskId)).build());
+            }
+            catch (NotFoundException e) {
+                asyncResponse.resume(Response.status(NOT_FOUND).entity("Could not find the requested taskInfo").build());
```

Comment:
does this work with the tests?

Reply:
Before, the tests were checking that tasks running on coordinator A could be retrieved via coordinator B. I added some new tests to check more exceptional cases:

* One test retrieves `/v1/taskInfo/20221102_075648_00000_8ybuj.9.0.0` (which parses as a TaskId, but it won't exist on the coordinator since it's fake). In the browser, this gave the "Could not find the requested taskInfo" message, but in the unit test I can only check that it returned a 404.
* The other test is that `/v1/taskInfo/invalid` (i.e. doesn't parse into a TaskId) returns a 404. This is because we only have the `/v1/taskInfo/{taskId}` endpoint setup, and `invalid` doesn't match that.